Better x86 CPU feature detection by rewriting all CPUID-related code

lrh2000 opened the pull request:
## Summary

This PR proposes the following two major changes:
 - Implement `arch/x86/cpu/cpuid.rs`, which replaces [the `cpuid` module from the `x86` crate](https://docs.rs/x86/latest/x86/cpuid/index.html).
   - In the future, we should not use the `cpuid` module from the `x86` crate anymore.
 - Implement `arch/x86/cpu/extension.rs`, which replaces [`arch::CPU_FEATURES`](https://github.com/asterinas/asterinas/blob/f56d2f3bad8ff769b50507356c0306488926d1cb/ostd/src/arch/x86/mod.rs#L53).
   - This mirrors the APIs of `arch/riscv/cpu/extension.rs`.

Since the generation of `/proc/cpuinfo` is closely related to the way to execute the CPUID instruction, it was also completely rewritten in this PR. I carefully examined the Linux implementation to ensure more accurate results than the previous implementation.

## Motivation

We previously detect x86 CPU features with the help of the `CPU_FEATURES` static variable:
https://github.com/asterinas/asterinas/blob/f56d2f3bad8ff769b50507356c0306488926d1cb/ostd/src/arch/x86/mod.rs#L53

This is far from being ideal due to the following reasons:
 - The [`FeatureInfo` type](https://docs.rs/x86/latest/x86/cpuid/struct.FeatureInfo.html) collects the feature information only from the CPUID 0x01 leaf, but there is other feature information from many other CPUID leaves. We can take a look at [the Linux implementation](https://github.com/torvalds/linux/blob/aaf724ed69264719550ec4f194d3ab17b886af9a/arch/x86/include/asm/cpufeatures.h):
```
/* Intel-defined CPU features, CPUID level 0x00000001 (EDX), word 0 */
/* Don't duplicate feature flags which are redundant with Intel! */
/* Intel-defined CPU features, CPUID level 0x00000001 (ECX), word 4 */
/* Intel-defined CPU features, CPUID level 0x00000007:0 (EBX), word 9 */
/* Intel-defined CPU features, CPUID level 0x00000007:1 (EAX), word 12 */
/* Intel-defined CPU features, CPUID level 0x00000007:0 (ECX), word 16 */
/* Intel-defined CPU features, CPUID level 0x00000007:0 (EDX), word 18 */

/* AMD-defined CPU features, CPUID level 0x80000001, word 1 */
/* More extended AMD flags: CPUID level 0x80000001, ECX, word 6 */
/* AMD-defined CPU features, CPUID level 0x80000008 (EBX), word 13 */
/* AMD SVM Feature Identification, CPUID level 0x8000000a (EDX), word 15 */
/* AMD-defined CPU features, CPUID level 0x80000007 (EBX), word 17 */
/* AMD-defined memory encryption features, CPUID level 0x8000001f (EAX), word 19 */
/* AMD-defined Extended Feature 2 EAX, CPUID level 0x80000021 (EAX), word 20 */
```
 - The `FeatureInfo` type is opaque to us, which prevents us from extending it. The upstream may not act quickly to add feature bits introduced in newer CPUs, so we have to wait for the upstream. For example, as of today, [`ExtendedFeatures`](https://docs.rs/x86/latest/x86/cpuid/struct.ExtendedFeatures.html) reports nothing about the "Indirect Branch Tracking (IBT)" support, nor about the "Protection Keys, Supervisor (PKS)" support.
 - The `FeatureInfo` type is opaque to us, which prevents us from optimizing it. In the future, we may want to remove the `Once` wrapper and use atomic variables to make the feature check faster, which requires us to break the type down.
 - The RISC-V APIs and the x86-64 APIs are inconsistent. The RISC-V implementation currently has `cpu::extension` module.

Meanwhile, we're in a real mess when it comes to choosing how to execute the CPUID instruction. We sometimes use `core::arch::x86_64::__cpuid`, sometimes use `x86::cpuid::cpuid!()`, and sometimes use `x86::cpuid::CpuId`:
https://github.com/asterinas/asterinas/blob/f56d2f3bad8ff769b50507356c0306488926d1cb/ostd/src/arch/x86/kernel/apic/xapic.rs#L63
https://github.com/asterinas/asterinas/blob/f56d2f3bad8ff769b50507356c0306488926d1cb/kernel/src/arch/x86/cpu.rs#L577
https://github.com/asterinas/asterinas/blob/f56d2f3bad8ff769b50507356c0306488926d1cb/kernel/src/arch/x86/cpu.rs#L300

It is also really easy to forget to check whether the CPUID leaf exists before trying to access it:
https://github.com/asterinas/asterinas/blob/f56d2f3bad8ff769b50507356c0306488926d1cb/kernel/src/arch/x86/cpu.rs#L577

We should check before accessing, like this:
https://github.com/asterinas/asterinas/blob/f56d2f3bad8ff769b50507356c0306488926d1cb/ostd/src/arch/x86/mod.rs#L164-L170

In short, the current implementation has too many problems, so a complete rewrite is necessary.

## APIs

Public APIs to execute the CPUID instruction:
```rust
// ostd/arch/x86/cpu/cpuid.rs

/// Executes the CPUID instruction for the given leaf and subleaf.
///
/// This method will return `None` if the leaf is not supported.
pub fn cpuid(leaf: u32, subleaf: u32) -> Option<CpuidResult>;
```

APIs to define extensions and check extensions:
```rust
// ostd/arch/x86/cpu/extension.rs

/// Checks if the specified set of ISA extensions are available.
pub fn has_extensions(required: IsaExtensions) -> bool;

define_isa_extensions! {
    leaf 1, subleaf 0 => {
        X2APIC,       Ecx(21), "The processor supports x2APIC feature.";
        TSC_DEADLINE, Ecx(24), "The processor's local APIC timer supports \
                                one-shot operation using a TSC deadline value.";
        // ...

        XAPIC,        Edx( 9), "APIC On-Chip.";
        // ...
    }

    leaf 7, subleaf 0 => {
        FSGSBASE,     Ebx( 0), "Supports RDFSBASE/RDGSBASE/WRFSBASE/WRGSBASE.";
        // ...
    }
}
```

## Fixed BUGs

I notice the following BUGs during the refactoring and fix them:
 - `CpuInfo` contains some per-CPU information like the core ID and the (initial) APIC ID, so it must be generated on each CPU. However, it seems that `aster-nix` does not current have a way to run an initialization routine on each processor, so this is currently implemented by triggering an IPI call:
```rust
        if !CPU_INFORMATION.get_on_cpu(CpuId::bsp()).is_completed() {
            // There can be races, but they do not matter at all.
            inter_processor_call(&CpuSet::new_full(), || {
                let guard = disable_preempt();
                CPU_INFORMATION
                    .get_on_cpu(guard.current_cpu())
                    .call_once(|| CpuInformation::new(&guard));
            });
        }
```
 - Some fields in `/proc/cpuinfo` (e.g., "physical id") are filled with wrong information (you can try to run `cat /proc/cpuinfo | grep 'physical id'` in Linux and you should be able to realize that our current implementation is just wrong):
https://github.com/asterinas/asterinas/blob/f56d2f3bad8ff769b50507356c0306488926d1cb/kernel/src/arch/x86/cpu.rs#L414-L418
 - We detect some CPU features by checking the wrong bits. For example, the `XAPIC` feature bit is the 9-th bit, not the 8-th bit:
https://github.com/asterinas/asterinas/blob/f56d2f3bad8ff769b50507356c0306488926d1cb/ostd/src/arch/x86/kernel/apic/xapic.rs#L62-L65
 - Maybe more...

tatetian reviewed (commented):
Bravos! This is great work. I really appreciate your effort.

lrh2000 commented:
> it could be at least useful for identifying grammar mistakes.

That's interesting, but it seems that our observations are not really the same. I saw Copilot reviews in other PRs, but they're very noisy and even the meaningful ones are not particularly insightful. I didn't realize that it is effective in fixing grammar issues either. What specific grammar mistakes were you referring to?

Meanwhile, I do think AI should be a great tool to identify and correct some style issues. For example,
 - quote the type name in `` ` ` ``;
 - write a summary line starting with the third-person singular verb for methods' doc comments;
 - fix short-hand conventions, e.g, `CPU` instead of `Cpu`/`cpu`, `vDSO` instead of `VDSO`/`vdso`.

These rules are hard to check by traditional tools but should be relatively simple for AIs to check if we write some suitable prompts. But I don't think we can guide the Github Copilot to do so. By default, the Github Copilot sometimes generates invalid reviews that suggest the wrong things (e.g., it said `/// Returns` should be `/// Return`).

lrh2000 commented:
Updates: https://github.com/asterinas/asterinas/compare/fd5e4b857492d65c45dd35ae2b997b02809634ae..2f4ff730fb44c1e4292bff395f329d47cd72c590

---

The cpuinfo GVisor test is still not enabled because it checks for the presence of the "bogomips" field, which we don't have. I am not sure if it is suitable to add the field and report a fake value (e.g., the TSC frequency?). So I leave it as it is now.

However, I've tested the implementation on my laptop and can confirm that all the fields except the following contain identical values to those of Linux:
 - "microcode" is not supported (it seems that we need to obtain that value from some MSR):
https://github.com/asterinas/asterinas/blob/e435c88e345f75cd2b9589e42e7ecf1755aec165/kernel/src/arch/x86/cpu.rs#L821
 - "cpu MHz" is not implemented correctly:
https://github.com/asterinas/asterinas/blob/e435c88e345f75cd2b9589e42e7ecf1755aec165/kernel/src/arch/x86/cpu.rs#L246-L248
 - "flags" is not fully implemented:
https://github.com/asterinas/asterinas/blob/e435c88e345f75cd2b9589e42e7ecf1755aec165/kernel/src/arch/x86/cpu.rs#L501-L502
 - "vmx flags"/"bugs"/"bogomips" is not supported:
https://github.com/asterinas/asterinas/blob/e435c88e345f75cd2b9589e42e7ecf1755aec165/kernel/src/arch/x86/cpu.rs#L862-L871

---

Additionally, I removed the IPI workaround and added support for `init_on_each_cpu` for `aster-nix`. I think this change will help https://github.com/asterinas/asterinas/pull/2398.

tatetian commented:
> What specific grammar mistakes were you referring to?

Maybe not grammar mistakes, but it did found some silly mistakes in my PRs that I should have noticed.

> Meanwhile, I do think AI should be a great tool to identify and correct some style issues. For example,
> 
> * quote the type name in `` ` ` ``;
> * write a summary line starting with the third-person singular verb for methods' doc comments;
> * fix short-hand conventions, e.g, `CPU` instead of `Cpu`/`cpu`, `vDSO` instead of `VDSO`/`vdso`.

Some standards or conventions for project-specific instructions to AI agents have emerged, e.g., [`Agents.md`](https://agents.md/) accepted by Gemini, Cursor, Github Copilot Agent, Zed, and Warp among many, `CLAUDE.md`, `Gemini.md`, and `.github/copilot-instructions.md` for Github Copilot (not agent). We can refer to our code style guide in these files. This will make Copilot and agents more useful to Asterinas.

> I saw Copilot reviews in other PRs, but they're very noisy and even the meaningful ones are not particularly insightful.

That's why I said Copilot is the dummest. But these undesirable outputs from Copilot may also result from an suboptimal configuration. Do you know that you can configure the AI model that Github Copilot uses? The default one is free, but it is not smart.

Anyway, despite that I share your dissatisfaction on Copilot, I found the false-positive results reported by Copilot are easy to dismiss with a minimal effort, yet the rest results are indeed useful. Plus, it is fun to watch how silly a coding model can be.

tatetian commented:
> In fact, it is not free. It requires Copilot Pro.

Copilot Pro is free to students 😄

tatetian reviewed (approved):
LGTM. Thanks for the contribution!

lrh2000 commented:
> Copilot Pro is free to students

Indeed, I haven't activated GitHub Education, so that must be why Copilot Review isn't available to me. I checked my account settings previously, but they said Copilot was already enabled.
